### DISCUSSION TITLE
Managing exceptions in C

### DISCUSSION THREAD

**Discussion by FUDCo** [Moddable-OpenSource/moddable#1331] [Q&A] [answered]
Background: I'm writing some C code to provide XS bindings for SQLite.

In the C-API doc (https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/tools/tools.md), with respect to `xsBeginHost` and `xsEndHost`, it says:

> Uncaught exceptions that occur between the calls the `xsBeginHost` and `xsEndHost` do not propagate beyond `xsEndHost`.

I further read the docs to say that if I'm in some random C function, in order to allocate temporary variables using the `xsVars` macro I'll need to enclose my code in `xsBeginHost` and `xsEndHost`.  The question is, what do I do if that code needs to throw a JS exception?

According to my reading, if I have a C function that needs to be able to BOTH allocate XS vars on the JS stack (using `xsVars`) AND throw a JS exception (using `xsThrow`), I appear to be up the creek without a paddle.  I suspect I'm reading this wrong, but my experimentation seems to bear this interpretation out.  I know it must be possible in principle to do this, since you _can_ do both of these things inside a C host function that is wired to JS with `xsNewHostFunction`.  However, if I have a helper C function that is not invoked directly from JS but rather from within the body of other C code (say, from several different host functions), I seem to be faced with a choice between allocating JS variables using `xsVars` or being able to throw exceptions.  I have a hard time believing this is really how things are, but I'm not sure what the real story is.

For the time being, I've worked around this by passing around the `xsMachine` context everywhere and placing intermediate JS values on the C stack in`xsSlot`-typed C variables.  This seems to work -- at least my code is passing all of my tests (of which there are quite a lot at this point) -- but I can't escape the feeling that I'm treading on very thin ice here since I'm pretty sure the values on the C stack are invisible the XS garbage collector.

Can anyone spare some enlightenment?

**Comment by phoddie** [comment]
First, you are correct that it is dangerous to put slots (as allocated by xsVars) on the C stack. It can be done with care, but it isn't a good general approach for exactly the reason you give - things may get garbage collected out from under you.

The part I'm not entirely following is "some random C function". It doesn't make sense for arbitrary C code to use the JavaScript stack or throw JavaScript exceptions. So, the functions must be a little less random than that!

There are two common categories of C functions that use the XS in C API: C functions called by JavaScript code and C functions that call JavaScript code. The former is a native implementation of a JavaScript method. The latter is usually some asynchronous trigger by a naive library, such as network data arriving or a timer callback firing.

- In the case of a C function called by JavaScript code, you don't need to use `xsBegin/EndHost` because XS sets up stack frame for you. You can allocate `xsVars` as needed and exceptions propagate in the usual JavaScript way.
- In the case of a C function that calls JavaScript code, you do need to use `xsBegin/EndHost` to establish a stack frame. Such a function is not invoked by JavaScript so it is the root of the JavaScript call stack - so it would be meaningless to propagate the exception beyond `xsEndHost`.

Note that a C function that is called by JavaScript code may also call JavaScript code. In that case, there is no need to set-up a new stack frame with `xsBegin/EndHost`. The called function (the receiver...) will establish its own stack frame. This is just like a JavaScript function calling another JavaScript function, and so exceptions propagate normally.

I hope this helps.

**Comment by FUDCo** [comment]
I apologize for being loose with my language when I spoke of "some random C function".  Let me give a more concrete example of the kind of thing that I'm talking about to see if I can make my quandary more clear.

In my case I'm writing C functions that will be called by JavaScript code, in service of implementing an API that wraps a library (in this case, SQLite) that is itself is written in C.  I understand I don't need to set up the stack for these functions.  However, various of the operations exposed by the API I'm implementing share a lot of common logic and hence a lot of common code in their implementation.  For example, a number of calls invoke SQL prepared statements and so share a common set of underlying mechanics for scanning parameters for type validation, translating the JavaScript parameter values into the appropriate corresponding C values, binding names for named prepared statement parameters, and so on.  It makes sense to factor out this parameter checking and translation code into a function of its own that is in turn called by the individual C functions that implement the individual JS methods.  However, each of those methods implements a slightly different operation from the others, which causes it to, for example, want a different slate of local variables using `xsVars`.  So method Foo might have 3 of them while method Bar has only 1 and method Baz has none at all.  However, after having invoked `xsVars(3)` or `xsVars(1)` or not having invoked `xsVars` at all, each wants to then call the joint parameter-checking-and-data-translation function, which needs to pick apart the JS parameters using the XS C API and needs a couple of local temp variables of its own as it's doing this.  If I use `xsBegin/EndHost` so I can then invoke, say, `xsVars(2)` I lose the ability to throw exceptions, which is needed since a key part of the job is parameter verification.  Alternatively I can put intermediate temp values into `xsSlot` C variables, but then I'm vulnerable to GC (actually, since the data structures being crawled are referenced on the JS call stack I think I don't really have to worry about them being reclaimed out from under me, but I suspect I still need to worry about objects being relocated).  Furthermore, the parameter checking code itself wants to be be factored into various internal functions that may call each other recursively, e.g., one to take apart parameters that are JS objects, another for JS arrays, another for primitive values, etc., and the same desiderata apply -- each may need temp variables such as you'd allocate with `xsVars` and each potentially needs to be able to throw exceptions.

Does this make sense?

**Reply by phoddie** [reply]
Thanks for the details. I understand better now. I still stand by everything I wrote above. ;)

Your shared native functions are a hybrid. They are C but they want to behave like a native implementation of a JavaScript function with their own stack frame and usual JavaScript function exception behavior.

You could just run with that -- make them JavaScript functions and call them with xsCall(). There would be a bit more overhead in the calls, but that doesn't seem likely to be significant here. I like this approach because it is straightforward by just following the usual JavaScript way. it would also give you the option to implement some of them in JavaScript, if that turns out to be useful.

> but I suspect I still need to worry about objects being relocated...

FWIW – objects don't relocate. Slots are fixed size in a heap that only contains slots, so there's never a need to compact. Chunks, on the other hand, do move (they are variable size, so there can be holes to fill). Strings and ArrayBuffers are chunks. And.... slots can contain chunk references.... and the GC updates those on compaction as needed, but it won't know to update any slots you've got on the native stack. But... really... you don't want to go down the road of trying to manage that. You'll eventually make a mistake that will take forever to track down.

**Reply by FUDCo** [reply]
> make them JavaScript functions and call them with xsCall()

I considered that, and it's an approach that might actually work for a couple of the _other_ things I'm grappling with, but in the specific example I cited these functions' job is essentially to translate JS data structures into C data structures that are ultimately going to be consumed by the underlying library I'm wrapping, so they really do need to be native code.

And yes, I absolutely don't want to take the path of attempting to second guess what the GC is doing and zigzag around it.  That seems too much like trying to cross a highway at night on foot -- you might get away with it just fine, but if you trip and fall...

A whacky hybridization approach which occurs to me: will it work for one host function to use `xsCall` to invoke another host function (assuming, of course, that the second function has been properly set up with `xsNewHostFunction` etc.)?  In other words, call one C function from another using the XS linkage?  A notable problem with this approach would be losing the ability to pass and return C pointers.  I supposed you could put any necessary pointery C context that needs to be shared into a separate place that could be referenced in parameters to these functions via a simple JS-passable identifier (such as an integer indexing into a global C array set up for this purpose -- and now I'm having FORTRAN flashbacks).  This seems like it could work, but also seems too clever by half, so I continue to hold out hope that something more straightforward is possible.

**Reply by phoddie** [reply]
Your "hybridization approach" is what I proposed above.

Maybe you are doing this upside down? It seems you have a big tree of JavaScript objects, some of which correspond to native objects for SQLite. Instead, maybe you should have JavaScript constructors for those native objects? You can then pass them to whatever API uses them. The native object would be stored in the host data, which is just an arbitrary pointer.

**Reply by phoddie** [reply]
@FUDCo – any further thoughts on this?

**Reply by FUDCo** [reply, marked as answer]
I ended up following a rather different course, which you might or might not approve of but I'll describe what I did in case it's useful.  [Note: one motivating complication is that I'm porting from C++ code written for the Node/V8 API to C code for XS, which is definitely an impedance mismatch, but I'm trying to preserve as much of the structure and logic of the original in order to minimize the time and effort required to port over any updates or bug fixes as they are released. However, the code I'm starting from is quite mature and stable so I'm not expecting a lot of that, rendering this marginally less insane].  In brief, the idea is to manage a portion of the XS variable stack myself, so that from XS' perspective all the internal C call graph descending from a JS function invocation looks like a single JS stack context.

I start by defining a context struct, which looks like this:
```c
typedef struct {
    xsMachine *the;
    int varStackSize;
    int sp;
} t_xsContext;
```
Host functions start with a preamble,
```c
   xsNewContext();
```
which is a macro defined as
```c
#define xsNewContext() t_xsContext ctx = initXSContext(the, STACKSIZE);
```
(where STACKSIZE is an empirically derived value; for my purposes I found a value of 10 to be perfectly adequate because my application has no recursion).

Most of the XS C API macros are predicated on having the `xsMachine` available in the variable `the`, but rather than following the pattern of passing `the` around everywhere, for everything descending the call graph beneath the host function (which is to say, the C entry point reached directly via a JS function/method call) I instead pass around one of my context structs.  The C functions that are internal to my library (which is most of them), instead of declaring `funcname(xsMachine *the, ...whatever`, I instead declare `funcname(t_xsContext ctx, ...whatever`.  (Note, by the way, that whereas the `xsMachine` is passed around as a pointer, the context struct is normally passed by value; this is important, as I'll explain at the end.)

These functions all start with a preamble:
```c
     xsPushContext();
```
which is another macro, defined as
```c
#define xsPushContext() xsMachine *the = ctx.the;
```

For the record, the `initXSContext` function used in the `xsNewContext` macro is defined as:
```c
t_xsContext initXSContext(xsMachine *the, int varStackSize) {
    t_xsContext result;
    result.the = the;
    result.varStackSize = varStackSize;
    result.sp = 0;
    xsVars(varStackSize);
    return result;
}
```

When I need a variable that's going to hold an `xsSlot`, I create it like this:
```c
    xsVarAlloc(foo);
```
with the macro definition:
```c
#define xsVarAlloc(name)  int name = allocateXSVar(&ctx);
```
the support function, `allocateXSVar`, is:
```c
int allocateXSVar(t_xsContext *ctxptr) {
    if (ctxptr->sp >= ctxptr->varStackSize) {
        throwError(*ctxptr, "xs var stack overflow");
    }
    return ctxptr->sp++;
}
```
Note that since the names defined by `xsVarAlloc` are simple integers, you can use them directly as arguments to `xsVar()`, much the way some of the code in the XS codebase does using `#define` constants, but with the names being block scoped rather than global.

This now allows me to write code like:
```c
t_databaseState *getDatabaseState(t_xsContext ctx, xsSlot *database) {
    xsPushContext();
    if (g_sqliteDatabaseInfo == NULL) {
        xsVarAlloc(s);
        xsVar(s) = xsGet(xsGlobal, xsID("SQLiteMetadata"));
        xsType type = xsTypeOf(xsVar(s));
        if (type == xsUndefinedType) {
            printf("no SQLiteMetadata object\n");
            throwError(ctx, "no SQLiteMetadata object found");
        }
        reestablishHostState(ctx, &xsVar(s));
    }
    xsVarAlloc(t);
    xsVar(t) = xsGet(*database, xsID("databaseID"));
    return g_sqliteDatabaseInfo[xsToInteger(xsVar(t))];
}
```
(no need to go into the details of what this code fragment is actually doing, just illustrating a use of the patterns)

This has allowed me to pretty directly translate ~2000 lines of C++ Node/V8 code to a corresponding ~1500 lines of C XS code, while having to do minimal violence to the logical structure of the former.

BTW, the reason the `ctx` struct is passed by value is so that the parameter simply initializes what amounts to a fresh context in the called function.  In particular, increments to the stack pointer get discarded when the function returns.  Any values remaining on the stack beyond the stack pointer will continue to be retained by XS (unless they get overwritten by a subsequent function call), and only made available for GC at the point where the XS host function at the top of the call tree returns, but I concluded the cost of that retention is a small price to pay for convenience, since the stack as a whole is small and short lived.

**Reply by phoddie** [reply]
Thanks for the detailed response. Very interesting stuff.

There are a couple of assumptions in this which are different from what I previously understood. You aren't using recursion, which definitely simplifies things. You are trying to maintain compatibility, to some degree, with existing APIs and implementation. That complicates things. ;)

Your solution seems reasonable given your constraints.

> Any values remaining on the stack beyond the stack pointer will continue to be retained by ...

The memory retention probably isn't a big deal in your environment. But, a freshly allocated variable in JavaScript is expected to be `undefined`. To maintain that, you could set the corresponding xsVar slots to `undefined `(`xsVar(1) = xsUndefined`) when rewinding your stack pointer, which would also prevent them from being retained. Or you could just set them to `undefined` in `allocateXSVar`.

> This has allowed me to pretty directly translate ~2000 lines of C++ Node/V8 code to a corresponding ~1500 lines of C XS code

That's pretty great.

> ... which you might or might not approve of...

Your solution is surely unusual but so are your requirements. You are entirely coloring inside the lines of the XS in C APIs, so it all seems quite reasonable and efficient.
